### ISSUE TITLE
http-outbound-probe-test failure with Node 8

### ISSUE THREAD

**Issue by gibfahn** [RuntimeTools/appmetrics#433]
Test failure with Node 8

#### Failure:

```
[Tue Apr 11 11:10:13 2017] com.ibm.diagnostics.healthcenter.mqtt INFO: Connecting to broker localhost:1883
    1..3
    # Subtest: HTTP Outbound Event
        ok 1 - Timestamp is an integer
        ok 2 - Should report GET as HTTP request method
        ok 3 - Should report http://localhost:8000/ as URL
        ok 4 - Should report world as value of hello header
        1..4
    ok 1 - HTTP Outbound Event # time=3.623ms

    not ok 2 - timeout!
      ---
      signal: SIGTERM
      handles:
        - type: Server
          events:
            - request
            - connection
          connectionKey: '6::::8000'
      ...
    not ok 3 - missing test
    # failed 2 of 3 tests
    # time=119892.613ms
not ok 3 - tests/probes/http-outbound-probe-test.js # time=120109.473ms
```

#### To reproduce:

Change the first URL for a different platform:

```bash
mkdir -p node8
# Use this line for Mac:
curl https://nodejs.org/download/nightly/v8.0.0-nightly20170410b470a85f07/node-v8.0.0-nightly20170410b470a85f07-darwin-x64.tar.gz | tar -xz --strip-components=1 -C "$PWD"/node8/
# For Linux use this line instead:
# curl https://nodejs.org/download/nightly/v8.0.0-nightly20170410b470a85f07/node-v8.0.0-nightly20170410b470a85f07-linux-x64.tar.gz | tar -xz --strip-components=1 -C "$PWD"/node8/
curl https://nodejs.org/download/nightly/v8.0.0-nightly20170410b470a85f07/node-v8.0.0-nightly20170410b470a85f07-headers.tar.gz > node8-headers.tar.gz
export PATH="$PWD/node8/bin:$PATH"
git clone --recursive https://github.com/RuntimeTools/appmetrics.git && cd appmetrics
git checkout 3.0.1
npm i --tarball=../node8-headers.tar.gz
npm test
```

**Comment by stalleyj** [comment]
I can now recreate this on the latest nightly build, continuing to investigate why our probe isn't working

**Comment by stalleyj** [comment]
I don't yet know why but this problem is due to the change in the way the module exports are done

**Comment by sam-github** [comment]
I think I see the problem.

```http.js
exports.get = function(...) {
   return exports.request(...);
}
exports.request = function(...) {
   // does its thing
```

appmetrics only patches `exports.request`: https://github.com/RuntimeTools/appmetrics/blob/c579f73fd7859d888e1f15e64bbfa86d307ebeaa/probes/http-outbound-probe.js#L36

This was fine, because all calls to `exports.get` called `exports.request`.

Now, node does:

```http.js
exports.get = get;
function get(...) {
   return request(...); // does not call via method on the `exports`
}
exports.request = request;
function request(...) {
   // does its thing
```
So even though `exports.request` is being replaced by a patch, `exports.get` is going straight to the original unpatched function.

TL;DR: you have to patch `exports.get` now. The code to do it should be identical, I suspect something like below will work, but I have not tried, I didn't see any tests.

The regression test should be easy, create a server with http, and make a request to it, observer expected event. You should be able to do entirely in a single node-process. If you do a `request()` and a `get()` both code paths will be hit and confirmed to work.

```
diff --git a/probes/http-outbound-probe.js b/probes/http-outbound-probe.js
index d65ae2d..3d575db 100644
--- a/probes/http-outbound-probe.js
+++ b/probes/http-outbound-probe.js
@@ -33,9 +33,9 @@ HttpOutboundProbe.prototype.attach = function(name, target) {
         if(target.__outboundProbeAttached__) return target;
         target.__outboundProbeAttached__ = true;
        
-        aspect.around(target, 'request',
-          // Before 'http.request' function
-          function(obj, methodName, methodArgs, probeData) {
+        aspect.around(target, 'request', patch);
+        aspect.around(target, 'get', patch);
+        function patch(obj, methodName, methodArgs, probeData) {
 
 
             // Get HTTP request method from options
```

**Comment by sam-github** [comment]
Addendum: Apologies! There are tests, they are right in the description :blush:. I'm pressed to prep for a LTS meeting, but I'll try to run them and check my theory/patch above.

**Comment by sam-github** [comment]
Btw, you probably want to make this one line `aspect.around(target, 'get', patch);` conditional on process.version being `semver.gt()` than 7, because it will result in the http.request looking like its nested inside an http.get on node < 8.

**Comment by stalleyj** [comment]
thanks @sam-github, so the exports.request is a copy of the original function rather than just pointing to it, this is the part I was missing. I have already made changes to the probe to patch both functions on node 8. Just finishing final testing.

**Comment by sam-github** [comment]
I think you meant to say the opposite, that exports.request *does* point to a function? And since in node 7, the function pointed to by http.get calls `function request()` indirectly, by calling whatever `exports.request` points to, its enough for appmetrics to assign its own wrapper function into `http.request`. In node 8, http.get calls `function request()` directly by value, so appmetrics doesn't have a way to make it call the wrapper.
